src/test/java/spoon/test/filters/FilterTest.java

```diff
@@ -466,4 +475,78 @@ public boolean matches(CtInvocation<?> element) {
 		}).get(0);
 		assertNotNull(invSize);
 	}
+	@Test
+	public void testQueryStepScannWithConsumer() throws Exception {
+		final Launcher launcher = new Launcher();
+		launcher.setArgs(new String[] {"--output-type", "nooutput" });
+		launcher.addInputResource("./src/test/java/spoon/test/filters/testclasses");
+		launcher.run();
+		
+		class Context {
+			int counter = 0;
+		}
+		Context context = new Context();
+		
+		QueryStep<CtClass<?>> l_qv = launcher.getFactory().getModel().getRootPackage().query().scan(new TypeFilter<>(CtClass.class));
+		
+		assertEquals(0, context.counter);
+		l_qv.forEach(cls->{
+			assertTrue(cls instanceof CtClass);
+			context.counter++;
+		});
+		assertTrue(context.counter>0);
+	}
+	
+	@Test
+	public void testQueryBuilderWithFilterChain() throws Exception {
+		final Launcher launcher = new Launcher();
+		launcher.setArgs(new String[] {"--output-type", "nooutput" });
+		launcher.addInputResource("./src/test/java/spoon/test/filters/testclasses");
+		launcher.run();
+		
+		class Context {
+			CtMethod<?> method;
+			int count = 0;
+		}
+		
+		Context context = new Context();
+
+		launcher.getFactory().Package().getRootPackage().query().scan(new TypeFilter<CtMethod<?>>(CtMethod.class))
+		.thenConsume((CtMethod<?> method) -> {context.method = method;})
```

Comment:
The problem is that we must visibly distinguish between 
A) then(...a function which produces elements for next step...)
B) thenConsume(... a function which just eats elements and produces nothing....)

We cannot let this distinction on the java compiler, because it would be confusing. Example:
`(element)->element.doSomething()`
Is that Consumer, which produces nothing or is it a function which produces result? Sometime the client wants to call function which returns values as consumer too.

If you write code like 
`...query().then(X).then(Y).then(Z)`
then it processes X and the result of X is sent to Y and result of **Y** is sent to Z

But if you write code like 
`...query().then(X).thenConsume(Y).then(Z)`
then it processes X and the result of X is sent to Y, which produces nothing and result of **X** is sent to Z

I am again open to change function name to anything better. May be I should learn some RxJava API and take over the names from that. Is there anybody who knows that API and can suggest names which are generally used and understandable?